### DISCUSSION TITLE
Clarify policy on migration

### DISCUSSION THREAD

**Discussion by sidkshatriya** [cloud-hypervisor/cloud-hypervisor#5914] [General]
In Cloud Hypervisor, migration from a running VM to another VM is supported if they are the same major version. Migration to a different version is not supported as I understand.

This feels like a major limitation. If we're running, for example, major version `N` there should always be a migration path to at least `N+1`.

Can there a clarification on what exactly is the policy of the project here ? QEMU for instance seems to go to heroic lengths to allow migration between different versions. Cloud Hypervisor should at least allow migration to the next version as a policy.

**Comment by sidkshatriya** [comment]
Clarification: Qemu has a concept of machine type which has a version associated with it. So you can do migration or save+restore on different Qemu versions as long as machine type remains the same.

E.g. qemu version **8.0** running machine type `virt-8.0` can do a migration to qemu version **8.1** running machine type `virt-8.0` . The problem here of course is that you are stuck on `virt-8.0` if you don't wish to shutdown the machine. It works similarly for qemu `savevm` and `loadvm`

Here I think firecracker provides more flexibility as later versions of firecracker can restore VMs created by an older firecracker and then you can conceivably perform a save of the VM on the new version of firecracker again thus providing an "upgrade path".

**Comment by rbradford** [comment]
We have all the infrastructure in place to support migration from older version to a newer version - however we don't currently guarantee that works.

**Reply by sidkshatriya** [reply]
> We have all the infrastructure in place to support migration from older version to a newer version - however we don't currently guarantee that works.

That's fantastic news. However without providing some sort of guarantee this feature can never be used in production. I would request that the project commit itself to this. Migration from older to newer versions is a very useful ability. At the minimum going from Cloud Hypervisor major version `N` to `N+1` should always work.

**Reply by liuw** [reply]
That's a fair point. The bare minimum is to support N to N+1 migration.

I'm not sure if that has been constantly tested in the CI system. We (upstream) will need to spend time to iron out the wrinkles first.

**Reply by sidkshatriya** [reply]
This is the main thing that is holding me back from using Cloud Hypervisor -- my particular usage scenario requires migration / save+restore be permitted between version N and N+1. Thanks for your response and I'll keep checking back for this feature...

**Comment by sidkshatriya** [comment]
Closing this discussion. When the project is ready to commit to N -> N+1 migration then an issue can be created.
